File: web/src/lib/panelSettings.ts
/**
 * Server-side memory of a tab's input fields — the same bargain the Simulation
 * tab has with the config.
 *
 * WHY (user 2026-09-07: "запоминай все последние настройки механических и
 * термических моделирований … всё одинаково для всех моделирований"): the
 * Mechanical and Thermal stores persisted their fields to `localStorage` only —
 * one browser's memory — and their `/last` restored just the parameters of the
 * last SOLVE.  A field set and not yet solved, or set in another browser, was
 * gone.  Now every persisted field is written to `/api/panel_settings/<panel>`
 * (debounced) and read back on the tab's first mount, server first; the
 * browser's copy stays the fallback for an offline page.
 *
 * Contract: the server stores the fields AS THE STORE KEEPS THEM (strings for
 * text fields), so the store's own migration is the single interpreter.
 */
const API = (import.meta.env.VITE_API_URL ?? 'http://localhost:8001') as string;
const BASE = `${API.replace(/\/$/, '')}/api/panel_settings`;

export type PanelName = 'mechanical' | 'thermal' | 'simulation' | 'mesh' | 'sweep';

export interface PanelSettings {
  settings: Record<string, unknown>;
  updated_at: string | null;
}

/** Read the caller's remembered fields; `null` when the server cannot answer
 *  (offline, not signed in) — never throws. */
export async function loadPanelSettings(panel: PanelName): Promise<PanelSettings | null> {
  try {
    const r = await fetch(`${BASE}/${panel}`, { cache: 'no-store' });
    if (!r.ok) return null;
    const j = await r.json() as { settings?: Record<string, unknown>; updated_at?: string | null };
    return { settings: j.settings ?? {}, updated_at: j.updated_at ?? null };
  } catch {
    return null;
  }
}

const _timers: Partial<Record<PanelName, number>> = {};
const _pending: Partial<Record<PanelName, Record<string, unknown>>> = {};

/** Queue a save of the panel's persisted fields; coalesced per panel so a
 *  field typed digit by digit is one PUT, not five.  The LAST snapshot wins. */
export function savePanelSettings(panel: PanelName, settings: Record<string, unknown>,
                                  delayMs = 700): void {
  _pending[panel] = settings;
  if (_timers[panel]) window.clearTimeout(_timers[panel]);
  _timers[panel] = window.setTimeout(() => {
    _timers[panel] = undefined;
    const body = _pending[panel];
    _pending[panel] = undefined;
    if (!body) return;
    void fetch(`${BASE}/${panel}`, {
      method: 'PUT', headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ settings: body }),
    }).catch(() => { /* offline: the browser copy is what we have */ });
  }, delayMs);
}

/** Pure: which server values to adopt.  Every key the panel persists that the
 *  server has is taken (server first — it is the memory shared by all
 *  browsers); keys the server never saw keep the store's current value. */
export function adoptSettings<T extends Record<string, unknown>>(
  current: T, server: Record<string, unknown> | null | undefined, keys: readonly (keyof T & string)[],
): Partial<T> {
  const out: Partial<T> = {};
  if (!server) return out;
  for (const k of keys) {
    if (!(k in server)) continue;
    const v = server[k];
    if (v === undefined) continue;
    // Same shape as the store keeps: a string field stays a string, a boolean
    // stays a boolean, an object (contacts) stays an object.
    const cur = current[k];
    if (cur !== undefined && cur !== null) {
      if (typeof v !== typeof cur) continue;
      // …and a LIST stays a list.  `typeof []` is 'object', so the line above
      // would happily drop an object where the panel keeps an array — which is
      // where the tabs' local comparison rows live since 2026-09-07, and a
      // `{}` adopted there is a table that throws on its first render.
      if (Array.isArray(cur) !== Array.isArray(v)) continue;
    }
    (out as Record<string, unknown>)[k] = v;
  }
  return out;
}

File: src/motor_ai_sim/routes/panel_settings.py
"""Server-side memory of a panel's INPUT fields — Mechanical, Thermal, and any
tab that comes after them.

WHY (user 2026-09-07: "запоминай все последние настройки механических и
термических моделирований … всё одинаково для всех моделирований"): the
Simulation tab keeps its operating point in the server config, so it comes back
on every browser, after every reload and every API restart.  The Mechanical and
Thermal tabs kept theirs in ``localStorage`` — one browser's memory — and their
``/last`` stored only the parameters of the last SOLVE.  A field the user set and
did not solve yet, or set in another browser, was gone.  This router is the
same bargain for every panel: the fields live here, the browser is a fallback.

One JSON file (``config/.panel_settings.json``, a dot-file like the ``.last_*``
stores — app state, not user configuration), keyed by panel and by the calling
user's e-mail (``shared`` when the caller is anonymous, i.e. auth not enforced).
Values are stored as the panel sends them (strings for the text fields), so the
panel's own migration/validation stays the single place that interprets them.
"""
from __future__ import annotations

import json
import os
import threading
import time
from typing import Any, Dict, Optional

from fastapi import APIRouter, Body, Header, HTTPException, Path

router = APIRouter(prefix="/api/panel_settings", tags=["panel-settings"])

_LOCK = threading.Lock()
_PANELS = ("mechanical", "thermal", "simulation", "mesh", "sweep")
_MAX_KEYS = 200
_MAX_BYTES = 64 * 1024


def _store_path() -> str:
    try:
        # Stage 1: the caller's WORKSPACE, which with none set is
        # ``Path(DEFAULT_CONFIG_PATH).parent`` — the old expression exactly.
        from motor_ai_sim.workspace import root as _ws_root
        base = str(_ws_root())
    except Exception:  # noqa: BLE001
        base = os.path.join(os.path.dirname(__file__), "..", "..", "..", "config")
    return os.path.abspath(os.path.join(base, ".panel_settings.json"))


def _load() -> Dict[str, Any]:
    p = _store_path()
    try:
        with open(p, encoding="utf-8") as f:
            d = json.load(f)
        return d if isinstance(d, dict) else {}
    except FileNotFoundError:
        return {}
    except Exception:  # noqa: BLE001 — a corrupt store is an empty store, never a 500
        return {}


def _save(d: Dict[str, Any]) -> None:
    p = _store_path()
    tmp = f"{p}.tmp{os.getpid()}"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(d, f, indent=1, ensure_ascii=False)
    os.replace(tmp, p)


def _who(authorization: Optional[str]) -> str:
    """The caller's e-mail, or ``shared`` when there is no signed-in user."""
    try:
        from motor_ai_sim.auth import resolve_user
        u = resolve_user(authorization)
        e = (u or {}).get("email")
        return str(e).strip().lower() if e else "shared"
    except Exception:  # noqa: BLE001
        return "shared"


def _check_panel(panel: str) -> str:
    if panel not in _PANELS:
        raise HTTPException(status_code=422, detail={
            "error": f"unknown panel '{panel}' — one of {', '.join(_PANELS)}",
            "invalid_parameters": ["panel"]})
    return panel


@router.get("/{panel}")
def get_panel_settings(panel: str = Path(...),
                       authorization: Optional[str] = Header(default=None)) -> Dict[str, Any]:
    """The remembered fields of one panel for the caller — ``{}`` when none.

    200 always: a tab that has never been used is the normal first state, not an
    error, and the panel falls back to its own defaults.
    """
    _check_panel(panel)
    who = _who(authorization)
    with _LOCK:
        d = _load()
    entry = ((d.get(panel) or {}).get(who)) or {}
    return {"panel": panel, "user": who, "settings": entry.get("settings") or {},
            "updated_at": entry.get("updated_at")}


@router.put("/{panel}")
def put_panel_settings(panel: str = Path(...),
                       body: Dict[str, Any] = Body(...),
                       authorization: Optional[str] = Header(default=None)) -> Dict[str, Any]:
    """Replace the caller's remembered fields of one panel.

    Body ``{"settings": {...}}`` — the panel's persisted fields as it keeps them
    (strings for text fields).  Only flat JSON scalars / small objects; bounded so
    a runaway client cannot turn the store into a dump.
    """
    _check_panel(panel)
    s = body.get("settings") if isinstance(body, dict) else None
    if not isinstance(s, dict):
        raise HTTPException(status_code=422, detail={
            "error": "body must be {\"settings\": {...}}", "invalid_parameters": ["settings"]})
    if len(s) > _MAX_KEYS:
        raise HTTPException(status_code=422, detail={
            "error": f"too many fields ({len(s)} > {_MAX_KEYS})", "invalid_parameters": ["settings"]})
    try:
        blob = json.dumps(s, ensure_ascii=False)
    except (TypeError, ValueError) as exc:
        raise HTTPException(status_code=422, detail={
            "error": f"settings are not JSON-serialisable: {exc}", "invalid_parameters": ["settings"]})
    if len(blob.encode("utf-8")) > _MAX_BYTES:
        raise HTTPException(status_code=422, detail={
            "error": f"settings too large ({len(blob)} bytes > {_MAX_BYTES})",
            "invalid_parameters": ["settings"]})
    who = _who(authorization)
    stamp = time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime()) + "+00:00"
    with _LOCK:
        d = _load()
        d.setdefault(panel, {})[who] = {"settings": s, "updated_at": stamp}
        _save(d)
    return {"panel": panel, "user": who, "settings": s, "updated_at": stamp}
